src/emulator/workQueue.ts

```diff
@@ -40,22 +40,25 @@ export class WorkQueue {
     }
 
     this.stopped = false;
-    this.interval = setInterval(() => {
-      if (this.shouldRunNext()) {
-        this.runNext();
+    while (!this.stopped) {
+      if (!this.queue.length) {
+        await new Promise((res) => {
+          this.notifyQueue = res;
+        });
+
+        const workPromise = this.runNext();
+        if (this.mode === FunctionsExecutionMode.SEQUENTIAL) {
+          await workPromise;
+        }
       }
-    }, 10);
+    }
   }
 
   /**
    * Stop processing work from the queue.
    */
   stop() {
     this.stopped = true;
-    if (this.interval) {
-      clearInterval(this.interval);
-      this.interval = undefined;
-    }
   }
 
   private shouldRunNext() {
```

Comment:
Please remove this.

Reply:
Removed but inline part of it into `submit()`

Reply:
Let's merge this discussion to the one I left on the test cases.